stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp

```diff
@@ -0,0 +1,212 @@
+ * This is equivalent to and faster than ordered_logistic_lpmf(y, x * beta,
+ * cuts).
+ *
+ * @tparam T_y type of integer vector of classes. It can be either
+ * `std::vector<int>` or `int`.
+ * @tparam T_x_scalar type of elements in the matrix of independent variables
+ * (features)
+ * @tparam T_x_rows compile-time number of rows of `x`. It can be either
+ * `Eigen::Dynamic` or 1.
+ * @tparam T_beta_scalar type of a scalar in the vector of weights
+ * @tparam T_cuts_scalar type of a scalar in the vector of cutpoints
+ * @param y a scalar or vector of classes. If it is a scalar it will be
+ * broadcast - used for all instances. Values should be between 1 and number of
+ * classes, including endpoints.
+ * @param x design matrix or row vector. If it is a row vector it will be
+ * broadcast - used for all instances.
+ * @param beta weight vector
+ * @param cuts cutpoints vector
+ * @return log probability
+ * @throw std::domain_error If any class is not between 1 and
+ * the number of cutpoints plus 2 or if the cutpoint vector is not sorted in
+ * ascending order or any input is not finite
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_y, typename T_x_scalar, int T_x_rows,
+          typename T_beta_scalar, typename T_cuts_scalar>
+typename stan::return_type<T_x_scalar, T_beta_scalar, T_cuts_scalar>::type
+ordered_logistic_glm_lpmf(
+    const T_y& y, const Eigen::Matrix<T_x_scalar, T_x_rows, Eigen::Dynamic>& x,
+    const Eigen::Matrix<T_beta_scalar, Eigen::Dynamic, 1>& beta,
+    const Eigen::Matrix<T_cuts_scalar, Eigen::Dynamic, 1>& cuts) {
+  using Eigen::Array;
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::VectorXd;
+  using std::exp;
+  using std::isfinite;
+
+  typedef typename partials_return_type<T_y, T_x_scalar, T_beta_scalar,
+                                        T_cuts_scalar>::type T_partials_return;
+  typedef typename std::conditional<T_x_rows == 1, double,
+                                    Array<double, Dynamic, 1>>::type T_location;
+
+  static const char* function = "ordered_logistic_glm_lpmf";
+
+  const size_t N_instances = T_x_rows == 1 ? length(y) : x.rows();
+  const size_t N_attributes = x.cols();
+  const size_t N_classes = length(cuts) + 1;
+
+  if (is_vector<T_y>::value && T_x_rows != 1) {
+    check_consistent_size(function, "Vector of dependent variables", y,
+                          N_instances);
+  }
+  check_consistent_size(function, "Weight vector", beta, N_attributes);
+  check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
+  check_ordered(function, "Cut-points", cuts);
+  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+  check_finite(function, "First cut-point", cuts[0]);
+
+  if (size_zero(y, cuts))
+    return 0;
+
+  if (!include_summand<propto, T_x_scalar, T_beta_scalar, T_cuts_scalar>::value)
+    return 0;
+
+  T_partials_return logp(0);
```

Comment:
The general preference is to have the variable declaration as close as possible to the variable's use